src/Control/Carrier/ContainerRegistryApi/Authorization.hs

```diff
@@ -0,0 +1,264 @@
+module Control.Carrier.ContainerRegistryApi.Authorization (
+  mkRequest,
+  getAuthToken,
+  applyBearAuth,
+
+  -- * for testing
+  RegistryAuthChallenge (..),
+  parseAuthChallenge,
+) where
+
+import Control.Algebra (Has)
+import Control.Carrier.ContainerRegistryApi.Common (
+  RegistryCtx,
+  acceptsContentType,
+  fromResponse,
+  getHeaderValue,
+  getToken,
+  logHttp,
+  originalReqUri,
+  updateToken,
+ )
+import Control.Carrier.ContainerRegistryApi.Errors (
+  ContainerRegistryApiErrorBody,
+  FailedToParseAuthChallenge (FailedToParseAuthChallenge),
+  UnknownApiError (UnknownApiError),
+ )
+import Control.Effect.Diagnostics (Diagnostics, fatal, fatalText)
+import Control.Effect.Lift (Lift, sendIO)
+import Control.Effect.Reader (Reader, ask)
+import Data.Aeson (FromJSON (parseJSON), decode', eitherDecode, withObject, (.:))
+import Data.ByteString.Lazy qualified as ByteStringLazy
+import Data.Map (Map)
+import Data.Map qualified as Map
+import Data.String.Conversion (encodeUtf8, toString, toText)
+import Data.Text (Text)
+import Data.Text qualified as Text
+import Data.Void (Void)
+import Effect.Logger (Logger)
+import Network.HTTP.Client (
+  Manager,
+  Request (method),
+  Response (responseBody, responseHeaders, responseStatus),
+  applyBasicAuth,
+  applyBearerAuth,
+  parseRequest,
+ )
+import Network.HTTP.Types (statusCode)
+import Network.HTTP.Types.Header (hWWWAuthenticate)
+import Text.Megaparsec (
+  MonadParsec (takeWhileP),
+  Parsec,
+  between,
+  chunk,
+  errorBundlePretty,
+  parse,
+  sepBy,
+  some,
+  (<|>),
+ )
+import Text.Megaparsec.Char (alphaNumChar, char)
+
+type Parser = Parsec Void Text
+
+-- | Request with registry authorization middleware.
+mkRequest ::
+  ( Has (Lift IO) sig m
+  , Has Diagnostics sig m
+  , Has Logger sig m
+  , Has (Reader RegistryCtx) sig m
+  ) =>
+  Manager -> -- Request Manager to use
+  Maybe (Text, Text) -> -- Credentials to use when retrieving authorization token
+  Maybe [Text] -> -- Content-Type to request
+  Request -> -- Request to make
+  m (Response ByteStringLazy.ByteString)
+mkRequest manager registryCred accepts req = do
+  token <- getToken =<< ask
+  token' <- getAuthToken registryCred req manager token
+  logHttp (applyContentType accepts $ applyBearAuth token' req) manager
+  where
+    applyContentType :: Maybe [Text] -> Request -> Request
+    applyContentType c r = case c of
+      Nothing -> req
+      Just c' -> r `acceptsContentType` (Text.intercalate ", " c')
+
+-- | Adds 'Authorization: Bearer <>', if 'Just token', otherwise id.
+applyBearAuth :: Maybe Text -> Request -> Request
+applyBearAuth t r = case t of
+  Nothing -> r
+  Just t' -> applyBearerAuth (encodeUtf8 t') r
+
+-- | Generates Auth Token For Request.
+--
+-- Refer to:
+--
+--  Docker Registry Spec: https://docs.docker.com/registry/spec/auth/token/
+--  OCI Spec:
```

Comment:
Remove Comment - as OCI does not specify authorization workflow spec, but it is the same as Docker Registry one, after testing against commercial registeries